### PR TITLE
Adds in Overloading of APCs

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Adds in overloading of APCs--inspired by Goon.

Basically, if APCs have in excess of 5,000,000 watts available to them, they will start arcing and shocking mobs in range.

This will only shock a single random mob when it does proc and do so for 5-25 burn damage---insulated gloves obviously protect against this effect; chance of this occurring is 5% (per APC), every 2 seconds.

Map Related
- All engineering and atmospheric APCs (including in the power substation, solars, and turbine) are "shock proof", meaning, even if the grid is overloaded, they will not arc out electricity.

Other Map Related
- Removes Christmas Tree and Wrapping paper from cargo

Admins:
- Inducing this to happen, intentionally, for no reason at all is considered an act of Grand Sabotage and is generally something that non-antags shouldn't be doing except in a few very very niche cases.

### CODE REVIEW

**Comment by Crazylemon64** [comment]
FINALLY, the twerps who hotwire the tesla will get their comeuppance...

**Comment by FalseIncarnate** [comment]
Should drain a moderate amount of power from the grid / internal cell when it zaps. Seeing as this is a discharge due to excess, it would make sense to actually discharge available energy and not magically suck it all back into the grid for round 2+.

**Comment by Tastyfish** [comment]
Finally, the engineers will hopefully stop hotwiring the tesla. :+1:

**Comment by Fox-McCloud** [comment]
>Should drain a moderate amount of power from the grid / internal cell when it zaps. Seeing as this is a discharge due to excess, it would make sense to actually discharge available energy and not magically suck it all back into the grid for round 2+.

Wouldn't really do jack, as the energy is being generated by an external power source--it wouldn't make any sense for that to be impacted--that's really not how the electric grid works, anyway (there's a lot of magic in powernets).

File: code/modules/power/apc.dm [hunk 1]
```diff
@@ -1252,6 +1254,18 @@
 			charging = 0
 			chargecount = 0
 
+		if(excess >= 5000000 && !shock_proof) //If there's more than 5,000,000 watts in the grid, start arcing and shocking people.
+			if(prob(5))
+				var/list/shock_mobs = list()
+				for(var/C in view(get_turf(src), 5)) //We only want to shock a single random mob in range, not every one.
+					if(iscarbon(C))
+						shock_mobs +=C
+				if(shock_mobs.len)
```
Comment (Iamgoofball) [inline code/modules/power/apc.dm]: >ports tesla
>hardcodes a shock effect
>doesn't use tesla_shock()

**Comment by pinatacolada** [comment]
The space law page should be updated, I can foresee people not knowing and or arguing it's not sabotage

File: code/modules/power/apc.dm (hunk 1 above)
Comment (Fox-McCloud) [inline code/modules/power/apc.dm]: its' a fine proc, but I didn't want it bouncing around, zapping machines, causing EMPs, and the likes.

**Comment by Fox-McCloud** [comment]
>The space law page should be updated, I can foresee people not knowing and or arguing it's not sabotage

Of course! If this gets merged, that'll have to be updated for clarity.

File: code/modules/power/apc.dm (hunk 1 above)
Comment (Iamgoofball) [inline code/modules/power/apc.dm]: no fun allowed

**Comment by NTSAM** [comment]
wait does this mean I can help the welfare of the station by feeding off of somewhat overloaded APCs

i like

i likey alot

**Comment by Robertha89** [comment]
I am not sure how I feel about this, on one hand, I love hot wiring the grid, on the other hand... No, this pr just prevents me from hot wiring the tesla, unless i putt special effort into making sure it produces less then 5 million... and 5 million is so little and so easy to do, make it 500 million, it takes effort, at least 2 PA's hacked to 3 to do, instead of something you can output with a few upgraded SMES, this way you can still hotwire the engine, because power is great, and wires get cut often, and power is removed for reasons often, so getting it back fast is top priority, not to forget that you can't really play the game as most roles without power.

**Comment by Regen1** [comment]
Well that's a whole lot of selfishness.
The pro's does not counter the cons of hotwiring the engine, and it is already a borderline thing to do, which will get you arrested.
Adding a new feature to it, not only stopping people from pumping the ridiculous amount of 500 MILLION FUCKING W into the powernet, it also opens for traitors to do this more often.

File: code/modules/power/apc.dm [hunk 2]
```diff
@@ -1252,6 +1254,18 @@
...
+		if(excess >= 5000000 && !shock_proof) //If there's more than 5,000,000 watts in the grid, start arcing and shocking people.
+			if(prob(5))
+				var/list/shock_mobs = list()
+				for(var/C in view(get_turf(src), 5)) //We only want to shock a single random mob in range, not every one.
+					if(iscarbon(C))
+						shock_mobs +=C
+				if(shock_mobs.len)
+					var/mob/living/carbon/S = pick(shock_mobs)
+					S.electrocute_act(rand(5,25), "electrical arc")
+					playsound(get_turf(S), 'sound/effects/eleczap.ogg', 75, 1)
```
Comment (tigercat2000) [inline code/modules/power/apc.dm]: so this does play a sound
good

**Comment by Plotron** [comment]
As if exploding wires weren't enough.

Well, I never liked the super hot powernet anyway... Will give the CE more of a reason not to hotwire.

**Comment by Fox-McCloud** [comment]
@Robertha89 

I can understand the dislike for it, and to a degree, I don't like it either, butttt utlimately:

- Powergloves out and out dust people at this level
- touching electrified things is instant death at this level

These are both pretty insane things that need some obvious drawback to them.

This doesn't prevent you from getting the engine to insane levels, by any stretch--as I stated in the OP, all engineering APCs will never arc electricity *ever* (unless replaced)---you're still free to push the engine as much as you want without any consequences....as long as the SMESes gate it.

**Comment by Robertha89** [comment]
... WAIT, to be shock proof, because there are only 156 APC's in total on the map, and 18 of those will be shock proof, leaving 138 to be changed...

**Comment by Fox-McCloud** [comment]
@Robertha89 

What?

**Comment by Robertha89** [comment]
Oh, excuse me.
Would it be possible to also add a way to remove and add shock proofing to APC's

**Comment by Fox-McCloud** [comment]
>Would it be possible to also add a way to remove and add shock proofing to APC's 

Rebuilding the APC

**Comment by Robertha89** [comment]
So shock proof is basically on by default on every new APC then?

**Comment by Fox-McCloud** [comment]
>So shock proof is basically on by default on every new APC then?

No, the only APCs that are shock proof are pre-mapped ones; there's no way for players to make shock proof APCs themseslves.

**Comment by TheDZD** [comment]
[Relevant](https://youtu.be/e_DqV1xdf-Y)

**Comment by tigercat2000** [comment]
ooh merged
i have to go make a webm of this now

**Comment by Deanthelis** [comment]
No more wrapping paper?
But how will Tater Buzz give Security delightful presents full of horrible chemical death if he can't add a little red bow?

**Comment by NTSAM** [comment]
You can still get it from the Arts'n'craft vendor I believe.

**Comment by Deanthelis** [comment]
Oh, good.
Glad that's settled.
I have no problems with anything else in this PR.

**Comment by Iamgoofball** [comment]
:-1: hotwiring the tesla engine into the grid is the standard op procedure

( :+1: just please make sure there's a way to shockproof APCs and/or override it as an antag somehow)

**Comment by Crazylemon64** [comment]
Err, wasn't the wrapping paper there normally? (@ the miscellaneous map edit)

**Comment by Fox-McCloud** [comment]
@Crazylemon64 

Nope; I added it for Christmas time (wrapping paper, not the brown paper wrap).

**Comment by tigercat2000** [comment]
@Fox-McCloud cargo actually normally had 'gift' wrapping paper in their office, as it's the only way to wrap humans in straightjackets

**Comment by Fox-McCloud** [comment]
Peculiar, because, when I added the Christmas tree, there was absolutely none to be found, so I added some, explicitly for Christmas.

**Comment by tigercat2000** [comment]
might have ended up removed at one of the fifty cargo redesigns, but i do recall there being some

**Comment by Robertha89** [comment]
Correct, it was normally never used due to the many if's and butts, and the fact that there are like 3 straightjackets ingame. But it used to be one.

**Comment by Deanthelis** [comment]
I also don't really see how removing gift wrap was in the scope of a PR about APC overloads.

**Comment by Iamgoofball** [comment]
stop whining about it in this PR and re-add it yourselves, it takes about a minute at most

**Comment by Deanthelis** [comment]
@Iamgoofball I'm just wondering what qualifies as within scope or not, as I'm clumsily trying to learn how to do all of this Git stuff properly by observing other PR's. It's not whining when it's a genuine lack of understanding.

**Comment by tigercat2000** [comment]
@Deanthelis eh it's whatever, misunderstanding/lack of information, PR's being atomic really doesn't need to be enforced for minor fixes

**Pull request merged**